RecoJets/JetProducers/plugins/CSJetProducer.cc

```diff
@@ -0,0 +1,188 @@
+#include "FWCore/Framework/interface/MakerMacros.h"
+#include "RecoJets/JetProducers/plugins/CSJetProducer.h"
+#include "FWCore/Utilities/interface/Exception.h"
+#include "RecoJets/JetProducers/interface/JetSpecific.h"
+
+using namespace std;
+using namespace reco;
+using namespace edm;
+using namespace cms;
+
+CSJetProducer::CSJetProducer(edm::ParameterSet const& conf):
+  VirtualJetProducer( conf ),
+  csRho_EtaMax_(-1.0),
+  csRParam_(-1.0),
+  csAlpha_(0.)
+{
+  //get eta range, rho and rhom map
+  etaToken_ = consumes<std::vector<double>>(conf.getParameter<edm::InputTag>( "etaMap" ));
+  rhoToken_ = consumes<std::vector<double>>(conf.getParameter<edm::InputTag>( "rho" ));
+  rhomToken_ = consumes<std::vector<double>>(conf.getParameter<edm::InputTag>( "rhom" ));
+  csAlpha_ = conf.getParameter<double>("csAlpha");
+}
+
+void CSJetProducer::produce( edm::Event & iEvent, const edm::EventSetup & iSetup )
+{
+  // use the default production from one collection
+  VirtualJetProducer::produce( iEvent, iSetup );
+  //use runAlgorithm of this class
+
+
+  // fjClusterSeq_ retains quite a lot of memory - about 1 to 7Mb at 200 pileup
+  // depending on the exact configuration; and there are 24 FastjetJetProducers in the
+  // sequence so this adds up to about 60 Mb. It's allocated every time runAlgorithm
+  // is called, so safe to delete here.
+  fjClusterSeq_.reset();
+}
+
+//______________________________________________________________________________
+void CSJetProducer::runAlgorithm( edm::Event & iEvent, edm::EventSetup const& iSetup)
+{
+  // run algorithm
+  if ( !doAreaFastjet_ && !doRhoFastjet_) {
+    fjClusterSeq_ = ClusterSequencePtr( new fastjet::ClusterSequence( fjInputs_, *fjJetDefinition_ ) );
+  } else if (voronoiRfact_ <= 0) {
+    fjClusterSeq_ = ClusterSequencePtr( new fastjet::ClusterSequenceArea( fjInputs_, *fjJetDefinition_ , *fjAreaDefinition_ ) );
+  } else {
+    fjClusterSeq_ = ClusterSequencePtr( new fastjet::ClusterSequenceVoronoiArea( fjInputs_, *fjJetDefinition_ , fastjet::VoronoiAreaSpec(voronoiRfact_) ) );
+  }
+
+  fjJets_.clear();
+  std::vector<fastjet::PseudoJet> tempJets = fastjet::sorted_by_pt(fjClusterSeq_->inclusive_jets(jetPtMin_));
+
+  //Get local rho and rhomo map
+  edm::Handle<std::vector<double>> etaRanges;
+  edm::Handle<std::vector<double>> rhoRanges;
+  edm::Handle<std::vector<double>> rhomRanges;
+  
+  iEvent.getByToken(etaToken_, etaRanges);
+  iEvent.getByToken(rhoToken_, rhoRanges);
+  iEvent.getByToken(rhomToken_, rhomRanges);
+
+  //Starting from here re-implementation of constituent subtraction
+  //source: http://fastjet.hepforge.org/svn/contrib/contribs/ConstituentSubtractor/tags/1.0.0/ConstituentSubtractor.cc
+  //some minor modifications with respect to original
+  //main change: eta-dependent rho within the jet
+  for ( std::vector<fastjet::PseudoJet>::const_iterator ijet = tempJets.begin(), ijetEnd = tempJets.end(); ijet != ijetEnd; ++ijet ) {
```

Comment:
Do you mean the C++11 loop (  for(fastjet::PseudoJet ijet : tempJets )  ), or something else? I implemented now the C++11 loop. If this is currently the guideline I can also implement this for the other loops over vectors in this producer.